src/crypto/OlmDevice.js

```diff
@@ -764,67 +806,53 @@ OlmDevice.prototype.getOutboundGroupSessionKey = function(sessionId) {
  *     this session to us (normally empty).
  */
 
-/**
- * store an InboundGroupSession in the session store
- *
- * @param {string} senderCurve25519Key
- * @param {string} sessionId
- * @param {InboundGroupSessionData} sessionData
- * @private
- */
-OlmDevice.prototype._saveInboundGroupSession = function(
-    senderCurve25519Key, sessionId, sessionData,
-) {
-    this._sessionStore.storeEndToEndInboundGroupSession(
-        senderCurve25519Key, sessionId, JSON.stringify(sessionData),
-    );
+OlmDevice.prototype._unpickleInboundGroupSession = function(sessionData, func) {
+    const session = new Olm.InboundGroupSession();
+    try {
+        session.unpickle(this._pickleKey, sessionData.session);
+        return func(session);
+    } finally {
+        session.free();
+    }
 };
 
 /**
- * extract an InboundGroupSession from the session store and call the given function
+ * extract an InboundGroupSession from the crypto store and call the given function
  *
  * @param {string} roomId
  * @param {string} senderKey
  * @param {string} sessionId
+ * @param {*} txn Opaque transaction object from cryptoStore.doTxn()
  * @param {function(Olm.InboundGroupSession, InboundGroupSessionData): T} func
  *   function to call.
  *
- * @return {null} the sessionId is unknown
- *
- * @return {T} result of func
- *
  * @private
  * @template {T}
  */
 OlmDevice.prototype._getInboundGroupSession = function(
-    roomId, senderKey, sessionId, func,
+    roomId, senderKey, sessionId, txn, func,
 ) {
-    let r = this._sessionStore.getEndToEndInboundGroupSession(
-        senderKey, sessionId,
-    );
-
-    if (r === null) {
-        return null;
-    }
-
-    r = JSON.parse(r);
+    this._cryptoStore.getEndToEndInboundGroupSession(
+        senderKey, sessionId, txn, (sessionData) => {
+            if (sessionData === null) {
+                func(null);
+                return;
+            }
 
-    // check that the room id matches the original one for the session. This stops
-    // the HS pretending a message was targeting a different room.
-    if (roomId !== r.room_id) {
-        throw new Error(
-            "Mismatched room_id for inbound group session (expected " + r.room_id +
-                ", was " + roomId + ")",
-        );
-    }
+            // if we were given a room ID, check that the it matches the original one for the session. This stops
+            // the HS pretending a message was targeting a different room.
+            if (roomId !== null && roomId !== sessionData.room_id) {
```

Comment:
why do we now check for null?

Reply:
ah, because as part of the migration we ended up calling this not knowing (or caring) what room the session was for, so null skips the check. I totally neglected to document it - now fixed.